2.0 keyboard navigation mode

Saulis opened the pull request:
Connects to #487 

Support for Page Down, Page Up, Home, End etc. will be added in a separate PR. This can be merged safely before that.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-grid/653)
<!-- Reviewable:end -->

manolo commented:
Review status: 0 of 23 files reviewed at latest revision, 3 unresolved discussions.

---

*[index.html, line 19 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafBbBHAhnJqpzvhmlt:-KafBbBIEkD3i7RrKYbA:b7w2bdk) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/index.html#L19)):*
> ```HTML
> <body>
>   <style>
>     /* snatched from GitHub */
> ```

At first I saw this out of context, may be a comment would clarify what `kbd` means.
Otherwise this could be moved to the JS doc comment, html should be allowed there.

---

*[vaadin-grid-active-item-behavior.html, line 46 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KadLdOK3lbq1YwWsqoM:-KadLdOK3lbq1YwWsqoN:b1cj760) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-active-item-behavior.html#L46)):*
> ```HTML
> 
>       e.stopPropagation();
>       e.stopImmediatePropagation();
> ```

is it necessary both? I guess stopInmediatePropagation should cover stopPropagation, so this is redundant

---

*[vaadin-grid-cell-click-behavior.html, line 34 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KadQ89a4jVT_vdgVOBA:-KadQ89a4jVT_vdgVOBB:b6tlbk9) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-cell-click-behavior.html#L34)):*
> ```HTML
>     _onClick: function(e) {
>       // this event is only supposed to be fired from cells.
>       if (this.is !== 'vaadin-grid-sorter') {
> ```

why don't use localName here? `is` might not work in 2.0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 23 files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-grid-table-keyboard-behavior.html, line 11 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafSE0vWCP64rtOW-eu:-KafSE0wk2mPGjSPL9K_:bg5g81e) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-keyboard-behavior.html#L11)):*
> ```HTML
> 
>       :host([navigating]) [focused] > tr[focused] > [focused] ::content > .cell-content {
>         position: relative;
> ```

Rather make all "cell-content" relative by default (in the common styles) than having one of them relative based on a status

---

*[vaadin-grid-table-keyboard-behavior.html, line 14 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafNliogGeL8-1XZmds:-KafNlipZxbG4ZmNewOc:b-wp7wcq) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-keyboard-behavior.html#L14)):*
> ```HTML
>       }
> 
>       :host([navigating]) [focused] > tr[focused] > [focused] ::content > .cell-content::before {
> ```

This should be doable (without affecting the cell-content positioning or adding a new element) by using inset box-shadow for the focused cell-content directly. For example `box-shadow: inset 0 0 0 3px rgba(0,0,0,0.3);`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 23 files reviewed at latest revision, 9 unresolved discussions.

---

*[vaadin-grid-focusable-cell-container-behavior.html, line 30 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kafe53EDJoMCpG3ZjtK:-Kafe53EDJoMCpG3ZjtL:b-ax38og) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-focusable-cell-container-behavior.html#L30)):*
> ```HTML
>           this._focusedRow = row;
>           this._focusedCellIndex = Math.min(cellIndex, row.children.length - 1);
>           this._focusedCell = row.children[this._focusedCellIndex];
> ```

this._focusedCellIndex -> cellIndex

Also below in the forEach loop

---

*[vaadin-grid-table.html, line 209 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kafhh8oJkq9qM8a9_9X:-Kafhh8p_kB9LFvLRmxp:b-4x4c6h) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table.html#L209)):*
> ```HTML
>     <style include="vaadin-grid-table-styles"></style>
> 
>     <style include="vaadin-grid-active-item-themability-styles"></style>
> ```

Changing the order of style module includes was probably unintentional. That's what's breaking one of the themability tests.

---

*[vaadin-grid.html, line 181 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafiraA0qDSnCrRbdFJ:-KafiraA0qDSnCrRbdFK:bucxi2a) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid.html#L181)):*
> ```HTML
>       row.expanded = this._isExpanded(item);
>       row.toggleAttribute('active', item !== null && item == this.activeItem);
>       row.focused = row.index === this.$.scroller.$.items._focusedRowIndex;
> ```

The purpose of both of the 2 bottom lines is to set an attribute to the row. It's a bit strange they use different API's for the same purpose. Maybe we could unify them by having row.active reflect to attribute also.

---

*[test/themability.html, line 115 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KaffK6nVqZmTjLmUZQy:-KaffK6nVqZmTjLmUZQz:b-tvrz7c) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/test/themability.html#L115)):*
> last frozen overrides active

It shouldn't. That's what the test verifies.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Having this strange issue with native shadow enabled: After navigating in the header, click the body. While the virtual focus shifts to the body correctly, arrow keys are now scrolling the document.

---

Review status: 0 of 23 files reviewed at latest revision, 10 unresolved discussions.

---

*[vaadin-grid-table-scroll-behavior.html, line 155 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafltnopogCFFRpMrma:-Kafltnp1eLiyQ0tu3Ld:b977af7) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-scroll-behavior.html#L155)):*
> ```HTML
>       var elementRect = element.getBoundingClientRect();
> 
>       // Vertical
> ```

The "vertical" part should be skipped if the virtual focus isn't in the table body.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KafmA8HdWYrX72EWdba:bpqi8rk)*
<!-- Sent from Reviewable.io -->

Saulis commented:
Oh fudge. Let's see.

---

Review status: 0 of 23 files reviewed at latest revision, 11 unresolved discussions.

---

*[index.html, line 19 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafBbBHAhnJqpzvhmlt:-Kafm9EHXextier667HQ:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/index.html#L19)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

At first I saw this out of context, may be a comment would clarify what `kbd` means.
Otherwise this could be moved to the JS doc comment, html should be allowed there.
</blockquote></details>

Done.

---

*[vaadin-grid-active-item-behavior.html, line 46 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KadLdOK3lbq1YwWsqoM:-Kaflq1yr0X8QTKR29n7:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-active-item-behavior.html#L46)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

is it necessary both? I guess stopInmediatePropagation should cover stopPropagation, so this is redundant
</blockquote></details>

Done.

---

*[vaadin-grid-cell-click-behavior.html, line 34 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KadQ89a4jVT_vdgVOBA:-KafmL8HQjL9NBECFLU5:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-cell-click-behavior.html#L34)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

why don't use localName here? `is` might not work in 2.0
</blockquote></details>

Done.

---

*[vaadin-grid-focusable-cell-container-behavior.html, line 30 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kafe53EDJoMCpG3ZjtK:-KafmSi5x9VKDW9Lzql-:b-lx5q3a) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-focusable-cell-container-behavior.html#L30)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

this._focusedCellIndex -> cellIndex

Also below in the forEach loop
</blockquote></details>

It's actually done on purpose like this because of the `Math.min` call required to make sure we don't assign a cell out of range.

---

*[vaadin-grid-table-keyboard-behavior.html, line 11 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafSE0vWCP64rtOW-eu:-KafoXRyBfsLuZeLP97b:b-sjyqet) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-keyboard-behavior.html#L11)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Rather make all "cell-content" relative by default (in the common styles) than having one of them relative based on a status
</blockquote></details>

I guess we can remove the `position: relative` altogether now when `:before` isn't used anymore?

---

*[vaadin-grid-table-keyboard-behavior.html, line 14 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafNliogGeL8-1XZmds:-KafoWMmUvui4EfCkqsB:bxvyexq) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-keyboard-behavior.html#L14)):*
> box-shadow: inset 0 0 0 3px rgba(0,0,0,0.3);

Nice!

---

*[vaadin-grid-table-scroll-behavior.html, line 155 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafltnopogCFFRpMrma:-KafsRRAeCzCvgQbSSHu:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-scroll-behavior.html#L155)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

The "vertical" part should be skipped if the virtual focus isn't in the table body.
</blockquote></details>

Done.

---

*[vaadin-grid-table.html, line 209 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kafhh8oJkq9qM8a9_9X:-KafqBnD0PxpnYBwSDz6:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table.html#L209)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Changing the order of style module includes was probably unintentional. That's what's breaking one of the themability tests.
</blockquote></details>

Done.

---

*[vaadin-grid.html, line 181 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafiraA0qDSnCrRbdFJ:-KafpcGyy20GGHHciM5s:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid.html#L181)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

The purpose of both of the 2 bottom lines is to set an attribute to the row. It's a bit strange they use different API's for the same purpose. Maybe we could unify them by having row.active reflect to attribute also.
</blockquote></details>

Done.

---

*[test/themability.html, line 115 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KaffK6nVqZmTjLmUZQy:-Kafq-Jpjr2-4WGacRjf:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/test/themability.html#L115)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

> last frozen overrides active

It shouldn't. That's what the test verifies.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-Kafss9U9bcuakMOFHCi:b-qvka89)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-grid-table-keyboard-behavior.html, line 11 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KafSE0vWCP64rtOW-eu:-KafvFXR-DbhcrbbAA6x:b-wqg6x1) ([raw file](https://github.com/vaadin/vaadin-grid/blob/b220636bac34f51b0f94bb581da57272c39deb1b/vaadin-grid-table-keyboard-behavior.html#L11)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

I guess we can remove the `position: relative` altogether now when `:before` isn't used anymore?
</blockquote></details>

Yep, just commented in case we didn't switch to box-shadow. All good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 23 files reviewed at latest revision, 2 unresolved discussions.

---

*[demo/styling.html, line 234 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kag0M6iRbxrUflUXJk0:-Kag0M6jgIub5VmRy1JA:b-9gwv49) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/demo/styling.html#L234)):*
> ```HTML
>               };
> 
>               --vaadin-grid-focused-cell: {
> ```

I feel like the rounded demo should also include some code to style focused cells so as they are aligned with the demo styles. Now it looks ugly 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 23 files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-grid-table-keyboard-behavior.html, line 13 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAixaBhg_zEACLIjV:-KagAixbxsxC2rBV6qzd:bgub51b) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L13)):*
> ```HTML
>         box-shadow: inset 0 0 0 3px rgba(0, 0, 0, 0.3);
> 
>         @apply(--vaadin-grid-focused-cell);
> ```

Since this is a new mixin for the "cell-content", it could use a specificity test in themability.html

---

*[demo/styling.html, line 234 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kag0M6iRbxrUflUXJk0:-KagAPoGi0Uja1Ey3i-l:b-8o846y) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/demo/styling.html#L234)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I feel like the rounded demo should also include some code to style focused cells so as they are aligned with the demo styles. Now it looks ugly 
</blockquote></details>

Good catch. Styling the focused cell in "rounded rows" demo is difficult currently since the demo creates its own containers within the cells for styling. Unless we add "focused" property to the item instance model also, we're left with applying styles to the underlying cell.

I'd perhaps hide the focus style with `box-shadow: none;` for now.

Default focus style should also be hidden from MD demo since the spec doesn't define how a focused cell should look like.

That said, should the cell focus feature be opt-in?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 23 files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-grid-table-keyboard-behavior.html, line 89 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAs3d82qcQsVMeBgm:-KagAs3eRbpolNt1Cr4b:b-knp41m) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L89)):*
> ```HTML
>     },
> 
>     _focusNext: function(e) {
> ```

When entering in table body, _ensureVirtualFocusInViewport should be called

---

*[vaadin-grid.html, line 72 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kag6Liqa735L1e2SQNN:-Kag6Lir65a38Pxl226B:byzh2rt) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid.html#L72)):*
> ```HTML
> <kbd>Right</kbd> | Moves the keyboard focus to the next cell
> <kbd>Left</kbd> | Moves the keyboard focus to the previous cell
> <kbd>Space</kbd> | Activates the item on the focused body cell
> ```

I miss `next-page` `prev-page` `start` `end` keys. They are implemented in grid 1.x. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 3 of 19 files at r2, 2 of 13 files at r4, 3 of 10 files at r5.
Review status: 8 of 23 files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-grid-dynamic-columns-behavior.html, line 76 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagDsX_Bo8xWPwQrj2F:-KagDsX_Bo8xWPwQrj2G:bwt2jqk) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-dynamic-columns-behavior.html#L76)):*
> ```HTML
>         // to light children. We need to make sure footer focus trap is always
>         // the very last element that can be tabbed into.
>         if (Polymer.Settings.useNativeShadow) {
> ```

I would check if it's already the `lastChild` to avoid manipulating the DOM otherwise

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 8 of 23 files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-grid-table-keyboard-behavior.html, line 13 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAixaBhg_zEACLIjV:-KagTCz14Pq-nNCL0T93:bb6mt9b) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L13)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Since this is a new mixin for the "cell-content", it could use a specificity test in themability.html
</blockquote></details>

:+1:

---

*[demo/styling.html, line 234 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kag0M6iRbxrUflUXJk0:-KagS6UnPGVcR1ODkZcJ:b-rbwqbj) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/demo/styling.html#L234)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Good catch. Styling the focused cell in "rounded rows" demo is difficult currently since the demo creates its own containers within the cells for styling. Unless we add "focused" property to the item instance model also, we're left with applying styles to the underlying cell.

I'd perhaps hide the focus style with `box-shadow: none;` for now.

Default focus style should also be hidden from MD demo since the spec doesn't define how a focused cell should look like.

That said, should the cell focus feature be opt-in?
</blockquote></details>

I would fix these examples to teach the user how to deal with this mixin in these cases.
box-shadow to none is worst because the grid takes the focus and the user does not have any feedback, he can move the grid with the keyboard but does knows  what is happening.
I think a flag to enable/disable keyboard navigation would be a nice feature for some cases. Not sure if it should be in this PR, maybe ask PD @jouni 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 1 of 6 files at r1, 3 of 19 files at r2.
Review status: 12 of 23 files reviewed at latest revision, 8 unresolved discussions.

---

*[vaadin-grid-table-keyboard-behavior.html, line 212 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagWnbhzztp_J9hPnyY:-KagWnbhzztp_J9hPnyZ:bw4qe10) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L212)):*
> ```HTML
> 
>         this.fire('cell-activate', {model: this._virtualFocus._focusedCell.instance});
>       }
> ```

If the cell is in header  and the column is sortable should switch sorting

---

*[vaadin-grid.html, line 62 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagW4HmwOC52XmLjel8:-KagW4HmwOC52XmLjel9:ba4v4cx) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid.html#L62)):*
> ```HTML
> `--vaadin-grid-loading-backdrop` | Mixin applied to the loading backdrop | `{}`
> 
> ### Keyboard Navigation
> ```

This should be documented somehow in asciidocs at least in selection make a reference that the user can interact with the selection using space

---

*[test/.eslintrc.json, line 38 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagVAx59wQJd6OE8NS6:-KagVAx6HUhVdUG87GKJ:booziww) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/test/.eslintrc.json#L38)):*
> ```JSON
>     "defaultCellWidth": false,
>     "firefox": false,
>     "listenOnce": false,
> ```

I'd rather use `window.listenOnce` and avoid maintaining these set of methods here. The code should be clearer. 
Can be accomplished in a separated PD though

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 4 of 10 files at r5.
Review status: 16 of 23 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

Saulis commented:
Ok the native shadow scrolling issue is now fixed by moving the behavior back to grid.

---

Review status: 16 of 23 files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-grid-dynamic-columns-behavior.html, line 76 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagDsX_Bo8xWPwQrj2F:-Kahf-SoY3kS5G_E8lL7:btjn96b) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-dynamic-columns-behavior.html#L76)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I would check if it's already the `lastChild` to avoid manipulating the DOM otherwise
</blockquote></details>

Looks to me like browser does that check for us?

---

*[vaadin-grid-table-keyboard-behavior.html, line 13 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAixaBhg_zEACLIjV:-Kahgl8sEFbmuK3ccsSk:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L13)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Done.

---

*[vaadin-grid-table-keyboard-behavior.html, line 89 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAs3d82qcQsVMeBgm:-KahmJ7w7eoq-KXiUOdi:b-elppjh) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L89)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

When entering in table body, _ensureVirtualFocusInViewport should be called
</blockquote></details>

Well.. I'm not sure. I maybe wouldn't scroll before user explicitly wants to navigate using arrow keys. What says @tomivirkki ?

---

*[vaadin-grid-table-keyboard-behavior.html, line 212 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagWnbhzztp_J9hPnyY:-KahpspZRia6WaBo1eQ9:b-g677kd) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L212)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

If the cell is in header  and the column is sortable should switch sorting
</blockquote></details>

This could be handy, but as it's not part of the WAI ARIA specs (space is reserved only for selecting) and this isn't straight forward to implement I wouldn't do it at this point but rather separately.

---

*[vaadin-grid.html, line 72 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kag6Liqa735L1e2SQNN:-KahqORUq2ncFVr79sJo:b-4u949v) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid.html#L72)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I miss `next-page` `prev-page` `start` `end` keys. They are implemented in grid 1.x. 
</blockquote></details>

They are coming in a separate PR as mentioned in the desc.

---

*[demo/styling.html, line 234 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-Kag0M6iRbxrUflUXJk0:-KahrLfmIdXd2t3wUDhK:b-1rllnx) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/demo/styling.html#L234)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I would fix these examples to teach the user how to deal with this mixin in these cases.
box-shadow to none is worst because the grid takes the focus and the user does not have any feedback, he can move the grid with the keyboard but does knows  what is happening.
I think a flag to enable/disable keyboard navigation would be a nice feature for some cases. Not sure if it should be in this PR, maybe ask PD @jouni 
</blockquote></details>

I would keep at least some indication as default where the focus is. It's quite important, especially when tabbing through. As for enabling/disabling keyboard, I think we need a concrete use case for that.

Replaced the styles from MD and rounded demos.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KaheeIJzkDS_mccpfNt:b8ircfv)*
<!-- Sent from Reviewable.io -->

jouni commented:
Enter key should trigger sorting, like in 1.x. 

Keyboard nav is an accessibility feature, so I don’t think there should be a eay to disable it.

manolo commented:
> Enter key should trigger sorting, like in 1.x.

@jouni  What about prev-page/next-page/start/end  keys  ?

---

Review status: 13 of 23 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KaiN08iK5_HKMIVvNDJ:bzd3kol)*
<!-- Sent from Reviewable.io -->

jouni commented:
Couple of issues I noticed:

- The grid is not scrolled into view when it gets focus – if the grid is outside the viewport and you tab into it, the focused cell is not scrolled into viewport.
- The focused cell style is not removed when the grid loses focus (click outside the grid, the body for example)

jouni commented:
Also, the “silent focus mode” is not working in Safari. The focus outline is shown when the user clicks a cell. It should only become visible when the keyboard arrow/tab keys are used.

jouni commented:
In Firefox, there are unnecessary focus outlines around some elements (seems like the “focus trap” elements before and after the grid):
![](https://www.dropbox.com/s/1oohspcuyupvfb7/Screenshot%202017-01-18%2009.37.39.png?dl=0&raw=1)

manolo commented:
Review status: 13 of 23 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-grid-dynamic-columns-behavior.html, line 76 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagDsX_Bo8xWPwQrj2F:-KalH7ycSocLWjigYQf3:b-tyh028) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-dynamic-columns-behavior.html#L76)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Looks to me like browser does that check for us?
</blockquote></details>

Nice!, tested and browser does not detach/attach if the element is not moved

---

*[vaadin-grid-keyboard-navigation-behavior.html, line 89 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAs3d82qcQsVMeBgm:-KalHVaPqitbL0hI5BYU:b-mbn8ym) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-keyboard-navigation-behavior.html#L89)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Well.. I'm not sure. I maybe wouldn't scroll before user explicitly wants to navigate using arrow keys. What says @tomivirkki ?
</blockquote></details>

I think @jouni comment refers to this issue ?

`The grid is not scrolled into view when it gets focus – if the grid is outside the viewport and you tab into it, the focused cell is not scrolled into viewport.`

IMO, focusing something that is out of the viewport is confusing for the user

---

*[vaadin-grid-keyboard-navigation-behavior.html, line 212 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagWnbhzztp_J9hPnyY:-KalITMCe4_6M9Go_91O:bijq5vj) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-keyboard-navigation-behavior.html#L212)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

This could be handy, but as it's not part of the WAI ARIA specs (space is reserved only for selecting) and this isn't straight forward to implement I wouldn't do it at this point but rather separately.
</blockquote></details>

So it should be the Enter key like in FW grid.  @joun also commented `Enter key should trigger sorting, like in 1.x.`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 13 of 23 files reviewed at latest revision, 6 unresolved discussions.

---

*[demo/styling.html, line 70 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KalJkTeRfB5JHtconf2:-KalJkTeRfB5JHtconf3:bkr17vd) ([raw file](https://github.com/vaadin/vaadin-grid/blob/473b0ee21a17e37ace760f8b032b8fb9904a61e6/demo/styling.html#L70)):*
> ```HTML
> 
>               --vaadin-grid-focused-cell: {
>                 box-shadow: none;
> ```

Please add some indication here like in the rounded demo for instance `font-weight: bold;`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

Saulis commented:
OK, focused item should now be scrolled to, navigating mode should disable when user clicks outside grid, "silent focus" problem I couldn't reproduce anymore in Safari? The FF outline issue is also fixed.

---

Review status: 13 of 23 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-grid-keyboard-navigation-behavior.html, line 89 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagAs3d82qcQsVMeBgm:-KalQYsYYfXH7dlRHb2q:bymkt6m) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-keyboard-navigation-behavior.html#L89)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I think @jouni comment refers to this issue ?

`The grid is not scrolled into view when it gets focus – if the grid is outside the viewport and you tab into it, the focused cell is not scrolled into viewport.`

IMO, focusing something that is out of the viewport is confusing for the user

</blockquote></details>

Yeah, added the functionality.

---

*[vaadin-grid-keyboard-navigation-behavior.html, line 212 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagWnbhzztp_J9hPnyY:-KalR1gm0QBIG_jzy44b:b-limlva) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-keyboard-navigation-behavior.html#L212)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

So it should be the Enter key like in FW grid.  @joun also commented `Enter key should trigger sorting, like in 1.x.`
</blockquote></details>

Enter is reserved to go into interaction mode: https://www.w3.org/TR/wai-aria-practices-1.1/#gridNav_inside

---

*[demo/styling.html, line 70 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KalJkTeRfB5JHtconf2:-KalR04bPmg1rehKmrUl:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/473b0ee21a17e37ace760f8b032b8fb9904a61e6/demo/styling.html#L70)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Please add some indication here like in the rounded demo for instance `font-weight: bold;`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KalQ6MjJhD5JMVhBMqA:bo3m1bt)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 13 of 23 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-grid-keyboard-navigation-behavior.html, line 212 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagWnbhzztp_J9hPnyY:-KalSp23ETfzb8I7qGF0:bljzr0v) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-keyboard-navigation-behavior.html#L212)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Enter is reserved to go into interaction mode: https://www.w3.org/TR/wai-aria-practices-1.1/#gridNav_inside
</blockquote></details>

But sorting is an interaction, isn't it?, it might be press enter to focus the space button and space to shift and enter to exit from the edit mode.

Anyway we need a way for the user to sort the column using keyboard.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[test/themability.html, line 141 at r8](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KaliTxQDpf4QSmLlohc:-KaliTxQDpf4QSmLlohd:b4j3has) ([raw file](https://github.com/vaadin/vaadin-grid/blob/ef06e9cb93162532a351be972b2da2e64e715d95/test/themability.html#L141)):*
> ```HTML
>         grid._navigating = true;
> 
>         setColor('--vaadin-grid-focused-cell', color1);
> ```

The test should compare the focused cell mixin to some other mixin. Id say that focus mixin should be stronger than selected cell mixin. In that case the precondition should be:
```
setColor('--vaadin-grid-body-row-selected-cell', color1);
setColor('--vaadin-grid-focused-cell', color2);
```
...and then just check that the resulting color is actually color2

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

Saulis commented:
Review status: 13 of 23 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[test/themability.html, line 141 at r8](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KaliTxQDpf4QSmLlohc:-Kaltkm1_I2I6J6kshGl:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/ef06e9cb93162532a351be972b2da2e64e715d95/test/themability.html#L141)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

The test should compare the focused cell mixin to some other mixin. Id say that focus mixin should be stronger than selected cell mixin. In that case the precondition should be:
```
setColor('--vaadin-grid-body-row-selected-cell', color1);
setColor('--vaadin-grid-focused-cell', color2);
```
...and then just check that the resulting color is actually color2
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 6 files at r1, 6 of 19 files at r2, 4 of 13 files at r4, 4 of 11 files at r5, 1 of 9 files at r6, 3 of 8 files at r7, 5 of 6 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KamOg3ttyGD-CWFYHIK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

manolo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> awesome job guys ! :shipit: 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-grid-table-keyboard-behavior.html, line 212 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-KagWnbhzztp_J9hPnyY:-KapepF4Cw2qpmmRU7BD:b-7ojbbj) ([raw file](https://github.com/vaadin/vaadin-grid/blob/d406b83778828b711cb49d1530991b95c950a818/vaadin-grid-table-keyboard-behavior.html#L212)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

But sorting is an interaction, isn't it?, it might be press enter to focus the space button and space to shift and enter to exit from the edit mode.

Anyway we need a way for the user to sort the column using keyboard.
</blockquote></details>

So, lets not make this a blocker right now and leave for a separate PR, maybe after editing cells is done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-Kapf4YJi9WejGFb_uy0:b-5p5gld)*
<!-- Sent from Reviewable.io -->

jouni commented:
I have a blocker: hidden columns are not taken into account when navigating.

jouni commented:
I also kind of feel like the issue with not scrolling the grid into viewport when it gets focused is a blocker.

jouni commented:
@Saulis, what say you? This doesn’t lgtm yet.

Saulis commented:
@jouni fixed the hidden cells issue – not sure how to fix the scrolling issue since Chrome and FF scroll automatically when focusing an element inside the grid. (the focus traps).

Wondering if Safari needs some special CSS rule or something on focusable divs to make it scroll...

---

Review status: 21 of 23 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KaqlvNGFHS_HVUmpUca:bn4jd9i)*
<!-- Sent from Reviewable.io -->

Saulis commented:
OK, added a `&nbsp;` to focus traps to allow Safari scroll on into them.

---

Review status: 19 of 23 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-KaqxITv-MUM2vi8iN8_:b-jtr30l)*
<!-- Sent from Reviewable.io -->

limonte commented:
Works perfectly on Linux both in Chrome and Firefox  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 18 of 23 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/653#-:-Kav-iyL4U_MbBj6N4hl:bb3fsfq)*
<!-- Sent from Reviewable.io -->
